### DISCUSSION TITLE
rdRascalMCES timeout error

### DISCUSSION THREAD

**Discussion by yingycl** [rdkit/rdkit#7967] [Q&A]
I am using the rdRascalMCES from rdkit.Chem. I set the timeout parameter to be a large number or -1 (unlimited time) using 
opts = rdRascalMCES.RascalOptions()
opts.timeout = -1
But I always got **"Timed out after 61 seconds."** and the program did terminate quickly. Did I use the timeout parameter wrong or there is a bug in rdRascalMCES?

**Comment by DavidACosgrove** [comment]
@yingycl Can you please provide a snippet of code demonstrating how you are running RascalMCES, and also what version of the RDKit you are using?  I have just tried it in 2024.09.1 and it seems to be working ok for me.

**Reply by yingycl** [reply]
I am running it as follows:

from rdkit.Chem import rdRascalMCES
opts = rdRascalMCES.RascalOptions()
opts.similarityThreshold = 0.0
opts.timeout = -1
results = rdRascalMCES.FindMCES(mol_A, mol_B, opts)

where mol_A and mol_B are two molecules randomly selected from AIDS dataset (https://chrsmrrs.github.io/datasets/).

My rdkit version is '2024.03.5'

**Reply by yingycl** [reply]
I also tested rdkit version '2024.09.1' and encountered the same problem.

**Reply by DavidACosgrove** [reply]
Can you provide 2 example molecules in SMILES or SDF format? I don't understand the format of the datasets from that website.

with this notebook code
```
from rdkit import Chem
from rdkit.Chem import rdRascalMCES, rdDepictor, Draw
import time
rdDepictor.SetPreferCoordGen(True)
m1 = Chem.MolFromSmiles(r"O[C@@H]1CC[C@H](C[C@H]1OC)C[C@@H](C)[C@@H]4CC(=O)[C@H](C)/C=C(\C)[C@@H](O)[C@@H](OC)C(=O)[C@H](C)C[C@H](C)\C=C\C=C\C=C(/C)[C@@H](OC)C[C@@H]2CC[C@@H](C)[C@@](O)(O2)C(=O)C(=O)N3CCCC[C@H]3C(=O)O4")
m2 = Chem.MolFromSmiles(r"CCC1C=C(CC(CC(C2C(CC(C(O2)(C(=O)C(=O)N3CCCCC3C(=O)OC(C(C(CC1=O)O)C)C(=CC4CCC(C(C4)OC)O)C)O)C)OC)OC)C)C")
opts = rdRascalMCES.RascalOptions()
opts.maxBondMatchPairs = 2000
opts.timeout = 5
opts.similarityThreshold = 0.0
s = time.perf_counter()
res = rdRascalMCES.FindMCES(m1, m2, opts)
print(f"Time : {time.perf_counter() - s}")
res
print(f"{len(res[0].bondMatches())} : {res[0].bondMatches()}")
Draw.MolsToGridImage([m1, m2], subImgSize=(600,600))
```

I get

![image](https://github.com/user-attachments/assets/8fc60796-dc6f-466b-a367-8ab2f411d2e7)

With `opts.timeout = -1` I get
![image](https://github.com/user-attachments/assets/2d736e35-5d17-44ee-9cb8-0c263f252041)

**Reply by yingycl** [reply]
Thanks for the prompt reply. The provided example works fine on my side. I may have found how to trigger my problems. In my case, I use a loop to iteratively calculate the MCES for many pairs, and my code is written like

```
from rdkit.Chem import rdRascalMCES
dataset = XXX
for i in range(5):
     mol_A, mol_B = dataset[i]
     opts = rdRascalMCES.RascalOptions()
     opts.similarityThreshold = 0.0
     opts.timeout = -1
     results = rdRascalMCES.FindMCES(mol_A, mol_B, opts)
```

e.g. for dataset[1], the SMILES of two molecules are 
"Cn1cc(NC(=O)c2cc(NC(=O)Nc3cc(C(=O)Nc4cc(C(=O)Nc5ccc6cc(S(=O)(=O)O)cc(S(=O)(=O)O)c6c5)n(C)c4)n(C)n3)nn2C)cc1C(=O)Nc1ccc2cc(S(=O)(=O)O)cc(S(=O)(=O)O)c2c1.[NaH]"
"COc1cc(-c2ccc(N=Nc3ccc4c(S(=O)(=O)O)cc(S(=O)(=O)O)c(N)c4c3O)c(OC)c2)ccc1N=Nc1ccc2c(S(=O)(=O)O)cc(S(=O)(=O)O)c(N)c2c1O.[NaH]"
Then I got "Timed out after 61 seconds." as described before, although opts.timeout = -1. For the other molecules in my dataset, they are small molecules and rdRascalMCES works fine.

However, when I directly calculate the MCES of those two molecules outside a loop, rdRascalMCES also works normally and I get 
<img width="955" alt="image" src="https://github.com/user-attachments/assets/d11af178-c2c6-4a21-90d9-a2d9fdcde666">

So it seems that my problem only happens in loops when I repeatly call rdRascalMCES.

**Comment by DavidACosgrove** [comment]
@yingycl I still can't reproduce this behaviour.  Can you provide exactly the structures you are feeding in, please.  It might also be useful to know what OS you're using.
It's not relevant to this, I imagine, but it's probably worth taking the [NaH] part out of the molecules, and generally doing some normalisation and salt stripping first.  And also, you don't need to put the creation of the options object inside the loop - you can use the same one each time.  It will make the code tidier and negligibly more efficient.

**Reply by yingycl** [reply]
Thank you so much. I have finally found out what is going on in my program. I called rdRascalMCES somewhere else in my program, and do not reset the opts.timeout parameter. So it's not the problem in rdRascalMCES. It's my fault. I’m sorry for wasting your time.

Besides, I noticed that when using rdRascalMCES on some molecules, e.g. the following two, the program will crash. My OS is Ubuntu 20.04.6.

`m1 = Chem.MolFromSmiles(r"O=S(=O)(O)c1ccc2c(c1)NN(c1ccc(C=Cc3ccc(N=Nc4ccc(C=Cc5ccc(N6Nc7ccc(S(=O)(=O)O)cc7N6)cc5S(=O)(=O)O)c(S(=O)(=O)O)c4)cc3S(=O)(=O)O)c(S(=O)(=O)O)c1)N2.[NaH]")

m2 = Chem.MolFromSmiles(r"CCCCCCCCCCCCCCCCCCNc1ccn(C2OC(CO)C(OP(=O)(O)OCC3CCC(n4ccc(=N)[nH]c4=O)O3)C2O)c(=O)n1")`

**Reply by DavidACosgrove** [reply]
On my system, it doesn't crash, it is killed by the OS because it uses too much memory.  As a rough guide, the amount of resource used by this type of problem is exponential in the number of bonds that could match between the two molecules.  This pair of molecules is too large with too many bonds that are plausible matches, and so must be checked, but which aren't in fact part of a large MCES.  I don't think there's a solution in this case other than to use a similarity threshold that means it doesn't attempt to find the MCES in the first place.
